### ISSUE TITLE
Hide HttpRequest behind a custom context?

### ISSUE THREAD

**Issue by mgernand** [DXBrazil/kubeless-netcore-runtime#5]
Hi @allantargino,

I was paying around with Kubeless and this cool runtime for .NET Core 2.0 and I was thinking that it might be a good idea to hide the `HttpRequest` class from the function implementation to support creating a simple `netstandard2.0` classlib project and use a function handler class file (f.e. `Handler.cs*) and the `.csproj` for declaring dependencies, just as one would in Visual Studio or Visual Studio Code.

What do you think?

Cheers, Matt

**Comment by allantargino** [comment]
Hi @mgernand!

We're exactly thinking on this part of the project (https://github.com/kubeless/kubeless/issues/395)
We already fixed the liveness probe endpoint, and we're going to deliver the requirements part.

As mentioned in #4, a initial idea was to just receive the specific part of requirements, not the full .csproj:
```xml
<ItemGroup>
  <PackageReference Include="Microsoft.AspNetCore" Version="2.0.0" />
  <PackageReference Include="System.Net.NetworkInformation" Version="4.3.0" />
</ItemGroup>
```
And as soon as the container is up, the following command would run, creating a "dynamic" .csproj:
```shell
echo -e '<Project Sdk="Microsoft.NET.Sdk.Web">\n<PropertyGroup>\n<TargetFramework>netcoreapp2.0</TargetFramework>\n</PropertyGroup>\n'"$(cat requirements.xml)"'\n</Project>' > requirements.csproj | dotnet restore
```

But your idea seems so more practical. I think we should change the approach to your suggestion, since it's really practical generate a `.csproj` with `dotnet new ...`

**Comment by allantargino** [comment]
Yes, loved the idea to hide `HttpRequest` in a custom object. I really think that we still should embedded `HttpRequest` in this new object, since it has several useful information from the request - like Headers, QueryString, Path, HttpMethod and other.

**Comment by mgernand** [comment]
Yes, I think that might be a good idea.
I am already on it to implement this. Maybe I can come up with a pull request that implements some bits.

**Comment by allantargino** [comment]
You mean the requirements or the custom object?
Thanks!! We really open to received your PR :)

**Comment by allantargino** [comment]
Just to let you know, we are working on this separated repo just for convenience. As soon as we stabilizes the runtime, we can all work directly on Kubeless official repo

**Comment by mgernand** [comment]
> You mean the requirements or the custom object?

Both. ;-)

> Just to let you know, we are working on this separated repo just for convenience. As soon as we stabilizes the runtime, we can all work directly on Kubeless official repo

Yes, I wondered why the code is mirrored there. I'd like to help out in the Kubeless project also, but did not have the time to contact the guys there.

**Comment by allantargino** [comment]
That's great!! Thank you very much!!
This weekend I won't be able to code on this repo, but I can review any PRs you sent. (and on monday I plan to start coding here again, I'm finishing another contribution on other project)

**Comment by mgernand** [comment]
When we introduce a custom context class, this class must be known by the function class. So what would be a good solution for this? Provide a runtime specific NuGet package containing the context class? This would enable developers to build the function class in f.e. Visual Studio and they might run some unit-tests against their function.

**Comment by mgernand** [comment]
> Just to let you know, we are working on this separated repo just for convenience. As soon as we stabilizes the runtime, we can all work directly on Kubeless official repo

Do they have any plans to create git repo per runtime?

**Comment by mgernand** [comment]
I'm thinking that maybe it would be cool to support both types of functions:

- One that only receives a `dynamic` context instance with the content of the context (`HttpContext`, etc.)
- One that must implement an interface `IKubelessFunction` and receives a `IKubelessContext` context instance with the content.

What do you think?

**Comment by allantargino** [comment]
> When we introduce a custom context class, this class must be known by the function class. So what would be a good solution for this? Provide a runtime specific NuGet package containing the context class? This would enable developers to build the function class in f.e. Visual Studio and they might run some unit-tests against their function.

I agree with you, a nuget package would allow developers to do their own tests (as it is possible today) - but we could gain on the side of having a easier custom object than HttpRequest.

**Comment by allantargino** [comment]
> Do they have any plans to create git repo per runtime?
I don't know exactly their plans about it, we could talk to them when we sent a PR with these newest changes.

**Comment by mgernand** [comment]
>> Do they have any plans to create git repo per runtime?

> I don't know exactly their plans about it, we could talk to them when we sent a PR with these newest changes.

Sounds great. 👍
